File: src/StreamPair.cpp
//
// Created by Vicram on 8/31/2019.
//

#include <fstream>
#include <iostream>
#include <ios> // for ios_base::binary

#include "StreamPair.h"

/*
 * IMPLEMENTATION NOTES
 * Shared Pointer Constructors:
 *   For the cases where the filename is given, I'm using shared_ptr constructor 3:
 *   https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr
 *   This will do 'delete ptr' on the object, which is the desired behavior.
 * 
 *   For the cases where no filename is given, we don't want to do 'delete ptr'
 *   because std::cin and std::cout are global variables. So I instead use constructor 4
 *   and pass in a dummy deleter which doesn't do anything. I think that the way I created
 *   and passed the deleter is valid, because one of the examples in that ref page declares
 *   the deleter like this:
 *     [](auto p) {std::cout << "Call delete from lambda...\n"; delete p;}
 * 
 * Error Handling:
 *   This reference page documents all the ways that I/O functions can fail:
 *   https://en.cppreference.com/w/cpp/io/ios_base/iostate
 *   It looks like constructing ifstream/ofstream can only fail if the file cannot
 *   be opened. If that happens, then the failbit is set; see here (we're using constructor 4):
 *   https://en.cppreference.com/w/cpp/io/basic_ifstream/basic_ifstream
 *
 * Binary Mode:
 *   By default, it seems that ifstream/ofstream are opened in text mode. Details for
 *   text mode are here: https://en.cppreference.com/w/cpp/io/c/FILE#Binary_and_text_modes
 *   This doesn't matter for POSIX systems but it does for Windows. On Windows, CRLF
 *   will be converted to LF on input and vice versa for output.
 *   The docs for the constructors for basic_ifstream and basic_ofstream discuss how
 *   to add the mode arg. There is also some relevant documentation from Microsoft:
 *   https://docs.microsoft.com/en-us/cpp/standard-library/binary-output-files?view=msvc-160
 *
 *   NOTE: std::cin and std::cout are opened in text mode! And according to SO, there
 *   is not a platform-independent way to change that: https://stackoverflow.com/a/7587701
 *   In other words, on Windows this program does not function 100% correctly unless
 *   the user gives a filename for both input and output.
 */

auto deleter = [](auto){}; // parameter name omitted to silence "unused parameter" warnings

void StreamPair::check_in(const std::string& inputfile) {
    if (in->fail()) {
        std::cerr << "Failed to open input file \"" << inputfile << "\". Exiting now." << std::endl;
        throw FileError();
    }
}

void StreamPair::check_out(const std::string &outputfile) {
    if (out->fail()) {
        std::cerr << "Failed to open output file \"" << outputfile << "\". Exiting now." << std::endl;
        throw FileError();
    }
}

StreamPair::StreamPair(const std::string &inputfile, const std::string &outputfile) : 
    in(new std::ifstream(inputfile, std::ios_base::binary)),
    out(new std::ofstream(outputfile, std::ios_base::binary)) {
        check_in(inputfile);
        check_out(outputfile);
}

StreamPair::StreamPair(const std::string &inputfile, bool) : 
    in(new std::ifstream(inputfile, std::ios_base::binary)),
    out(&std::cout, deleter) {
        check_in(inputfile);
}

StreamPair::StreamPair(bool, const std::string &outputfile) :
    in(&std::cin, deleter),
    out(new std::ofstream(outputfile, std::ios_base::binary)) {
        check_out(outputfile);
}

StreamPair::StreamPair(bool, bool) :
    in(&std::cin, deleter),
    out(&std::cout, deleter) {}

File: java/StreamPair.java
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.charset.StandardCharsets;

/**
 * This class encapsulates the input and output streams.
 * Once you are finished with both streams, you should call closeStreams.
 */
class StreamPair {
    boolean usingStdin;
    InputStream in;
    boolean usingStdout;
    PrintStream out;

    /**
     * Given the input filename and output filename, this constructor will open
     * two streams and set the in and out members accordingly.
     * If either inputfile or outputfile is null, then stdin/stdout will be used
     * for the stream, respectively.
     * If we fail to open either file then this constructor will throw a
     * StreamPairException with an error message. The caller should print this
     * message to stderr before exiting.
     *
     * If this constructor throws an exception, you do not need to call closeStreams.
     *
     * Note: this constructor should only throw an IOException in the case that
     * we failed to open the output file, AND we had already opened an input file,
     * AND the input file threw an IOException upon close().
     */
    StreamPair(String inputfile, String outputfile) throws StreamPairException, IOException {
        if (inputfile == null) {
            usingStdin = true;
            in = System.in;
        } else {
            usingStdin = false;

            InputStream instream = null;
            try {
                Path path = Paths.get(inputfile);
                instream = Files.newInputStream(path);
            } catch (InvalidPathException | IOException | SecurityException e) {
                assert instream == null; // TODO is it possible for this to be non-null here?
                // We could print out a more detailed error message, but this is identical
                // to the message printed by the C++ implementation.
                throw new StreamPairException("Failed to open input file \"" + inputfile + "\". Exiting now.");
            }
            in = new BufferedInputStream(instream);
        }

        if (outputfile == null) {
            usingStdout = true;
            out = System.out;
        } else {
            usingStdin = false;

            OutputStream outstream = null;
            try {
                Path path = Paths.get(outputfile);
                outstream = Files.newOutputStream(path);
            } catch (InvalidPathException | IOException | SecurityException e) {
                assert outstream == null;
                // Because we may have opened an input file, we must also close it.
                if (!usingStdin) {
                    in.close();
                }
                throw new StreamPairException("Failed to open output file \"" + outputfile + "\". Exiting now.");
            }
            try {
                out = new PrintStream(outstream, false, StandardCharsets.US_ASCII.name());
            } catch (UnsupportedEncodingException e) {
                assert false; // US_ASCII is guaranteed to be supported by any implementation of the JVM.
            }
        }
    }

    /**
     * This method closes any streams associated with files.
     */
    void closeStreams() throws IOException {
        try {
            if (!usingStdin) {
                in.close();
            }
        } finally {
            if (!usingStdout) {
                out.close();
            }
        }
    }
}
